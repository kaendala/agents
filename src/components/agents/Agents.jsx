import "./Agents.scss";
import React, { useEffect, useState } from "react";
import axios from "axios";
import Card from "../common/card/Card";
import Select from "react-select";
import { formatCurrency } from "../../utils/number";
import { useHistory } from "react-router-dom";
import { useSelector } from "react-redux";

const Agents = () => {
  const history = useHistory();
  const [filterList, setFilterList] = useState(null);
  const [sort, setSort] = useState(null);
  const [currentList, setCurrentList] = useState(null);
  const [page, setPage] = useState(1);
  const [totalPage, setTotalPage] = useState();
  const rate = 10000;
  const inputValue = useSelector((state) => state.incomeReducer.income);

  const options = [
    { value: "name-asc", label: "Name (A-Z)" },
    { value: "name-desc", label: "Name (Z-A)" },
    { value: "income-desc", label: "Income: High first" },
    { value: "income-asc", label: "Income: Low first" },
  ];
  useEffect(() => {
    axios.get("/assets/json/AGENTS_LIST.json").then((res) => {
      if (res.data && res.data.length > 0) {
        const listFilt = res.data.filter((element) => {
          if (
            element.income <= inputValue + rate &&
            element.income >= inputValue - rate
          ) {
            return element;
          } else {
            return null;
          }
        });
        setFilterList(listFilt);
        setTotalPage(Math.ceil(listFilt.length / 3));
      }
    });
  }, [inputValue]);

  useEffect(() => {
    if (filterList && filterList.length > 0) {
      let list = filterList.filter((element, index) => {
        if (index + 1 <= page * 3) {
          return element;
        } else {
          return null;
        }
      });
      if (sort) {
        const type = sort.split("-")[0];
        const sortSelect = sort.split("-")[1];
        list = list.sort((a, b) => {
          if (sortSelect === "desc") return a[type] > b[type] ? -1 : 1;
          else {
            return b[type] > a[type] ? -1 : 1;
          }
        });
      }

      setCurrentList(list);
    } else {
      setCurrentList(null);
    }
  }, [page, filterList, sort]);

  const more = () => {
    let pageCurrent = page + 1;
    setPage(pageCurrent);
  };
  const less = () => {
    let pageCurrent = page - 1;
    setPage(pageCurrent);
  };

  const removeAgent = (agent) => {
    const id = agent.id;
    let list = filterList.filter((element) => {
      if (element.id !== id) {
        return element;
      } else {
        return null;
      }
    });
    setFilterList(list);
    const totalPage = Math.ceil(list.length / 3);
    setTotalPage(totalPage);
    if (page > totalPage) {
      setPage(totalPage);
    }
    saveAgent(agent);
  };
  const saveAgent = (agent) => {
    const savedAgents = JSON.parse(localStorage.getItem("agents"));
    if (savedAgents && savedAgents.length > 0) {
      savedAgents.push(agent);
      localStorage.setItem("agents", JSON.stringify(savedAgents));
    } else {
      localStorage.setItem("agents", JSON.stringify([agent]));
    }
  };

  return (
    <div className="agents">
      {currentList && currentList.length > 0 && (
        <div>
          <h1 className="title">Your matches</h1>
          <h2 className="subtitle">
            Your income:{" "}
            <strong>{formatCurrency("en-US", "USD", 0, inputValue)}</strong>
          </h2>
          <div className="cards">
            <div className="contentSelect">
              <h4 className="label">Order agents by</h4>
              <Select
                options={options}
                className="select"
                classNamePrefix="react-select"
                onChange={(e) => setSort(e.value)}
              />
            </div>
            <div className="cardsContainer">
              {currentList.map((agent) => (
                <Card key={agent.id} agent={agent} saveAgent={removeAgent} />
              ))}
            </div>
          </div>
          <div className="contentButtons">
            <button
              className="link"
              onClick={() => less()}
              disabled={page === 1}
            >
              <h3>Show less -</h3>
            </button>
            <button
              className="link"
              onClick={() => more()}
              disabled={page === totalPage}
            >
              <h3>Show more +</h3>
            </button>
          </div>
        </div>
      )}
      {!currentList && (
        <div>
          <h1 className="noAgents">
            “No available Agents based on your income. Please try a different
            income value.”
          </h1>
          <button className="link" onClick={() => history.push("/")}>
            <h3>&#8592; Go back </h3>
          </button>
        </div>
      )}
    </div>
  );
};

export default Agents;
