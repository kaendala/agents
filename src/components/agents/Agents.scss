@import "../../style/global.scss";
.agents {
  .title {
    margin-top: 38px;
  }
  .subtitle {
    margin: 12px 0 32px;
    color: $GD600;
  }
  .cards {
    width: 66.3%;
    margin: 0 auto;
    .contentSelect {
      text-align: left;
      .label {
        color: $GD400;
      }
      .select {
        outline: none;
        width: 24.52%;
        height: 46px;
        margin: 9px 0 15px;
        padding: 0;
        border-radius: 4px;
        border: solid 1px $GL300;
        background-color: $W0;
        display: flex;
        align-items: center;

        .react-select__control {
          height: 30px;
          padding: 0 1.2vw;
          width: 100%;
          border: none;
          box-shadow: none;
          &:hover {
            box-shadow: none;
          }
          .react-select__indicator-separator {
            display: none;
          }
        }
        .react-select__control--menu-is-open {
          background-color: $W0;
        }

        .react-select__menu {
          border-radius: 4px;
          background-color: $W0;
        }
        .react-select__option--is-selected {
          background-color: $B03;
        }
      }
    }
    .cardsContainer {
      display: flex;
      flex-wrap: wrap;
      justify-content: space-between;
    }
  }
  .contentButtons {
    width: 66.3%;
    margin: 0 auto 46px;
    text-align: right;
    .link {
      background-color: transparent;
      border: none;
      h3 {
        font-weight: 700;
        color: $B03;
      }
      &:disabled {
        h3 {
          color: $GD100;
        }
      }
    }
  }
}
