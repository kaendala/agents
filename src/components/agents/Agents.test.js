import React from 'react';
import { act,fireEvent,render } from '@testing-library/react';
import Agents from './Agents';
import configureMockStore from 'redux-mock-store';
import mockedAxios from 'axios';
import { Provider } from 'react-redux';
import agentsData from '../../../public/assets/json/AGENTS_LIST.json';

const mockStore = configureMockStore();
jest.mock('axios');
const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

describe('Test for Card component', () => {
    let component;
    const storeInitialState = mockStore({
      incomeReducer: {
        income: 20000
        }
    });
    beforeEach(async() => {
        mockedAxios.get.mockResolvedValue({ data: agentsData });
        await act(async () => {
            component = render(<Provider store={storeInitialState}><Agents></Agents></Provider>);
        })
    })
    test('should load component Agents ', () => {
        expect(component).toMatchSnapshot();
    });
    test('should get 6 agents ', async() => {
        let agents= component.container.querySelectorAll(".card")
        expect(agents.length).toBe(3);
        component.container.querySelectorAll(".link")[1].click()
        component.rerender(<Provider store={storeInitialState}><Agents></Agents></Provider>)
        agents= component.container.querySelectorAll(".card")
        expect(agents.length).toBe(6);
    });
    test('should get 3 agents ', async() => {
        component.container.querySelectorAll(".link")[1].click()
        component.container.querySelectorAll(".link")[0].click()
        component.rerender(<Provider store={storeInitialState}><Agents></Agents></Provider>)
        const agents = component.container.querySelectorAll(".card")
        agents[0].click()
        agents[2].click()
        expect(agents.length).toBe(3);
    });
    test('should sort ASC', async() => {
        const sort = component.container.querySelector(".select");
        sort.click();
        const select=component.container.querySelector('.react-select__dropdown-indicator')
        fireEvent.mouseDown(select, {button: 0 });
        const option = component.container.querySelector('.react-select__option')
        fireEvent.click(option);
        component.rerender(<Provider store={storeInitialState}><Agents></Agents></Provider>)
        const agents = component.container.querySelectorAll(".card")
        expect(agents.length).toBe(3);
    });
     test('should sort DESC', async() => {
        const sort = component.container.querySelector(".select");
        sort.click();
        const select=component.container.querySelector('.react-select__dropdown-indicator')
        fireEvent.mouseDown(select, {button: 0 });
        const option = component.container.querySelectorAll('.react-select__option')[1]
        fireEvent.click(option);
        component.rerender(<Provider store={storeInitialState}><Agents></Agents></Provider>)
        expect(component).toBeTruthy();
     });
     test('should call push', async() => {
        mockedAxios.get.mockResolvedValue({ data: {} });
        await act(async () => {
            component = render(<Provider store={storeInitialState}><Agents></Agents></Provider>);
        })
         component.container.querySelector(".link").click()
        expect(mockHistoryPush).toBeCalled();
     });
    
})