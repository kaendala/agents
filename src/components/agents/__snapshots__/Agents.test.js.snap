// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for Card component should load component Agents  1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="agents"
      >
        <div>
          <h1
            class="title"
          >
            Your matches
          </h1>
          <h2
            class="subtitle"
          >
            Your income:
             
            <strong>
              $20,000
            </strong>
          </h2>
          <div
            class="cards"
          >
            <div
              class="contentSelect"
            >
              <h4
                class="label"
              >
                Order agents by
              </h4>
              <div
                class="select css-2b097c-container"
              >
                <span
                  aria-atomic="false"
                  aria-live="polite"
                  aria-relevant="additions text"
                  class="css-1f43avz-a11yText-A11yText"
                />
                <div
                  class="react-select__control css-yk16xz-control"
                >
                  <div
                    class="react-select__value-container css-g1d714-ValueContainer"
                  >
                    <div
                      class="react-select__placeholder css-1wa3eu0-placeholder"
                    >
                      Select...
                    </div>
                    <div
                      class="css-b8ldur-Input"
                    >
                      <div
                        class="react-select__input"
                        style="display: inline-block;"
                      >
                        <input
                          aria-autocomplete="list"
                          autocapitalize="none"
                          autocomplete="off"
                          autocorrect="off"
                          id="react-select-2-input"
                          spellcheck="false"
                          style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                          tabindex="0"
                          type="text"
                          value=""
                        />
                        <div
                          style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    class="react-select__indicators css-1hb7zxy-IndicatorsContainer"
                  >
                    <span
                      class="react-select__indicator-separator css-1okebmr-indicatorSeparator"
                    />
                    <div
                      aria-hidden="true"
                      class="react-select__indicator react-select__dropdown-indicator css-tlfecz-indicatorContainer"
                    >
                      <svg
                        aria-hidden="true"
                        class="css-tj5bde-Svg"
                        focusable="false"
                        height="20"
                        viewBox="0 0 20 20"
                        width="20"
                      >
                        <path
                          d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="cardsContainer"
            >
              <div
                class="card"
              >
                <div
                  class="avatar"
                >
                  <img
                    alt="avatar"
                    src="/assets/images/man.png"
                  />
                </div>
                <div
                  class="description"
                >
                  <h2
                    class="name"
                  >
                    <strong>
                      Rand
                    </strong>
                  </h2>
                  <h3
                    class="id"
                  >
                    ID: 
                    33
                  </h3>
                  <div
                    class="income"
                  >
                    Income  
                    <strong>
                      $12,580
                    </strong>
                  </div>
                </div>
              </div>
              <div
                class="card"
              >
                <div
                  class="avatar"
                >
                  <img
                    alt="avatar"
                    src="/assets/images/female.png"
                  />
                </div>
                <div
                  class="description"
                >
                  <h2
                    class="name"
                  >
                    <strong>
                      Kathy
                    </strong>
                  </h2>
                  <h3
                    class="id"
                  >
                    ID: 
                    34
                  </h3>
                  <div
                    class="income"
                  >
                    Income  
                    <strong>
                      $10,604
                    </strong>
                  </div>
                </div>
              </div>
              <div
                class="card"
              >
                <div
                  class="avatar"
                >
                  <img
                    alt="avatar"
                    src="/assets/images/female.png"
                  />
                </div>
                <div
                  class="description"
                >
                  <h2
                    class="name"
                  >
                    <strong>
                      Susan
                    </strong>
                  </h2>
                  <h3
                    class="id"
                  >
                    ID: 
                    35
                  </h3>
                  <div
                    class="income"
                  >
                    Income  
                    <strong>
                      $13,601
                    </strong>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="contentButtons"
          >
            <button
              class="link"
              disabled=""
            >
              <h3>
                Show less -
              </h3>
            </button>
            <button
              class="link"
            >
              <h3>
                Show more +
              </h3>
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="agents"
    >
      <div>
        <h1
          class="title"
        >
          Your matches
        </h1>
        <h2
          class="subtitle"
        >
          Your income:
           
          <strong>
            $20,000
          </strong>
        </h2>
        <div
          class="cards"
        >
          <div
            class="contentSelect"
          >
            <h4
              class="label"
            >
              Order agents by
            </h4>
            <div
              class="select css-2b097c-container"
            >
              <span
                aria-atomic="false"
                aria-live="polite"
                aria-relevant="additions text"
                class="css-1f43avz-a11yText-A11yText"
              />
              <div
                class="react-select__control css-yk16xz-control"
              >
                <div
                  class="react-select__value-container css-g1d714-ValueContainer"
                >
                  <div
                    class="react-select__placeholder css-1wa3eu0-placeholder"
                  >
                    Select...
                  </div>
                  <div
                    class="css-b8ldur-Input"
                  >
                    <div
                      class="react-select__input"
                      style="display: inline-block;"
                    >
                      <input
                        aria-autocomplete="list"
                        autocapitalize="none"
                        autocomplete="off"
                        autocorrect="off"
                        id="react-select-2-input"
                        spellcheck="false"
                        style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                        tabindex="0"
                        type="text"
                        value=""
                      />
                      <div
                        style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                      />
                    </div>
                  </div>
                </div>
                <div
                  class="react-select__indicators css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class="react-select__indicator-separator css-1okebmr-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class="react-select__indicator react-select__dropdown-indicator css-tlfecz-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-tj5bde-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="cardsContainer"
          >
            <div
              class="card"
            >
              <div
                class="avatar"
              >
                <img
                  alt="avatar"
                  src="/assets/images/man.png"
                />
              </div>
              <div
                class="description"
              >
                <h2
                  class="name"
                >
                  <strong>
                    Rand
                  </strong>
                </h2>
                <h3
                  class="id"
                >
                  ID: 
                  33
                </h3>
                <div
                  class="income"
                >
                  Income  
                  <strong>
                    $12,580
                  </strong>
                </div>
              </div>
            </div>
            <div
              class="card"
            >
              <div
                class="avatar"
              >
                <img
                  alt="avatar"
                  src="/assets/images/female.png"
                />
              </div>
              <div
                class="description"
              >
                <h2
                  class="name"
                >
                  <strong>
                    Kathy
                  </strong>
                </h2>
                <h3
                  class="id"
                >
                  ID: 
                  34
                </h3>
                <div
                  class="income"
                >
                  Income  
                  <strong>
                    $10,604
                  </strong>
                </div>
              </div>
            </div>
            <div
              class="card"
            >
              <div
                class="avatar"
              >
                <img
                  alt="avatar"
                  src="/assets/images/female.png"
                />
              </div>
              <div
                class="description"
              >
                <h2
                  class="name"
                >
                  <strong>
                    Susan
                  </strong>
                </h2>
                <h3
                  class="id"
                >
                  ID: 
                  35
                </h3>
                <div
                  class="income"
                >
                  Income  
                  <strong>
                    $13,601
                  </strong>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="contentButtons"
        >
          <button
            class="link"
            disabled=""
          >
            <h3>
              Show less -
            </h3>
          </button>
          <button
            class="link"
          >
            <h3>
              Show more +
            </h3>
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
